### DISCUSSION TITLE
Reduce memory restrictions when returning arrays in Fortran

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1159] [Ideas]
> > > > > Maybe we can improve the way we handle it in the bind_c function and avoid the extra allocation by directly using the returned array.
> > > > 
> > > > 
> > > > The problem is that `allocatable`s are automatically deallocated once they go out of scope. So if we point at the allocatable then by the time we get back to c we will be pointing at freed memory which is obviously dangerous
> > > 
> > > 
> > > Maybe this rule doesn't apply with bind(c) functions. [here](https://www.ibm.com/docs/en/xl-fortran-aix/15.1.2?topic=29113-example-allocatable-pointer-arguments) we have an example with a returned `allocatable` array in a bind(c) function
> > 
> > In the example you are linking to the allocatable is `intent(out)` which is why it is not deallocated. However the only reason they can do this is because the C side is using `#include "ISO_Fortran_binding.h"`. This was introduced in [Fortran 2018] (https://fortranwiki.org/fortran/show/Fortran+2018). Pyccel usually follows the Fortran 2003 standard as Fortran compilers are very slow to be updated. It seems that support for ISO_Fortran_binding is now available for the GNU and intel compilers (see [here](https://fortranwiki.org/fortran/show/Fortran+2018+status) but not for PGI. So using this would make it impossible to use PGI to accelerate with pyccel.
> > 
> > If we don't use ISO_Fortran_binding then we can't return an array directly, instead we need to wrap it in a `c_ptr` which leads to the "result variable" going out of scope
> 
> I see, one solution is to provide both ways of doing it because we can generate them both, or we can do a more complicated one using the `save` attribute in a module variable
> somehting like
> ```fortran
> module dealloc_test
> 
>    use, intrinsic :: iso_c_binding, only : c_double, c_ptr, c_size_t, c_associated, c_loc
> 
>    implicit none
> 
>    real(c_double), allocatable, target, protected, save :: data_array(:) ! Named, protected target data object
> 
> contains
> 
>    function alloc(size) bind(C, name='alloc') result(rv)
> 
>       ! Argument list
>       integer(c_size_t), value, intent(in) :: size
>       ! Function result
>       type (c_ptr) :: rv
> 
>       ! Local variables
>       integer(c_size_t) :: i
> 
>       ! Handling elided for an existing allocation of data_array
>       allocate( data_array(size) )
> 
>       rv = c_loc(data_array)
> 
>       return
> 
>    end function
> 
>    subroutine dealloc(ptr) bind(C, name='dealloc')
> 
>       ! Argument list
>       type(c_ptr), value, intent(in) :: ptr
> 
>       ! Check if input pointer is associated with module variable
>       if ( c_associated(ptr, c_loc(data_array)) ) then
>          deallocate( data_array )
>       end if
> 
>       return
> 
>    end subroutine
> 
> end module
> ```
> 

I don't think using save or a module variable would be sufficient to fix the problem. If the function is called twice with different
arguments and the result is saved in 2 different variables then the second call will overwrite the contents of the first array.

Generating different code depending on the available standard is an interesting idea. I would be for this as it would simplify a few things, but it is completely out of the scope of this PR. We should open a discussion to talk about this

_Originally posted by @saidctb and @EmilyBourne in https://github.com/pyccel/pyccel/issues/1056

**Comment by saidctb** [comment]
Both my suggestions can be done in another PR, I think it's fine to work with what we have now for the moment.
For the use of the `save` attribute, you are right it only works with one array and one allocation, to make it general we need to have a dictionary to keep track of all the allocated arrays, so each time the allocate function is called we add a new array to the dictionary, and when we want to deallocate it we can remove it from the dictionary.

**Reply by EmilyBourne** [reply]
This sounds really complicated to me. It means we need to keep track of where an array was created, and we somehow need to pass this information to numpy as it is numpy that deallocates the memory not us

**Reply by saidctb** [reply]
I'm not sure about the details but I know that we can add the [__del__](https://docs.python.org/3.9/reference/datamodel.html#object.__del__) method to Python object, so it's probably doable with Numpy Object, we can tell Numpy to call the function `dealloc` that we defined

**Reply by EmilyBourne** [reply]
Yes but doesn't that mean that we would then have an array of `Object` (which has it's own `__del__` method) instead of an array of `float64`/`int32`/etc

**Reply by saidctb** [reply]
As for keeping track of the object, I don't see the issue, each NumPy array will have a pointer to the function that does the deallocation,  so if we have multiple `bind_c` files, each one has its own function that has access to saved module variable that has all the returned allocated arrays

**Reply by saidctb** [reply]
I'm not sure , you probably right, or maybe NumPy provides us with some function that we can use

**Reply by EmilyBourne** [reply]
In python it is not an issue (assuming we can set a `__del__` function which I don't think we can for numpy's types). However if the call which creates the array is in a translated file then we need to use a different `deallocate` function if we also need to remove it from a dictionary.

E.g:
```python
def single_return(n : int):
    a = np.ones(n) # translated with save or module variable
    return a

def f():
   b = single_return(3)
   c = single_return(5)
   d = np.ones(4)
   # deallocate(b) -> single_return_deallocate(b)
   # deallocate(c) -> single_return_deallocate(c)
   # deallocate(d) -> deallocate(d)
```

**Reply by saidctb** [reply]
I think the numpy array object has a reference to its `__del__` method saved within the object, so maybe we can change it to our `__del__` method

**Reply by EmilyBourne** [reply]
If we do that I would be worried that all arrays would call this. I suspect the `__del__` method is a class function not an object function but I have never looked into it. In any case this doesn't sound simpler than having 1 copy in the wrapper

**Reply by saidctb** [reply]
yes, it will be different , it will be something like
```python
def single_return_deallocate(a):
    fortran_dealloc(a)
    numpy_dealloc(a)
```

**Reply by saidctb** [reply]
it's an improvement, it's more efficient and uses less memory by a factor of two, but yes it's more complicated for sure

**Reply by EmilyBourne** [reply]
Yes. That is why we need to keep track of where/how each array was created so we know exactly which deallocate function to call

**Reply by saidctb** [reply]
we can always call the Fortran `dealloc` which will do nothing if the array doesn't belong to it, and then call the numpy `dealloc`afterwards

**Reply by EmilyBourne** [reply]
> uses less memory by a factor of two

For a fraction of a second anyway. I'm surprised the compiler can't optimise this out.

It's a shame we can't use [`MOVE_ALLOC`](https://gcc.gnu.org/onlinedocs/gfortran/MOVE_005fALLOC.html)

**Reply by EmilyBourne** [reply]
> we can always call the Fortran `dealloc` which will do nothing if the array doesn't belong to it, and then call the numpy `dealloc` afterwards

We can but this could be really inefficient. If we do this then we have to call the `dealloc` function for all functions in the scope (either in the module or imported) which return arrays. Depending on the size of the project that could be a huge number of functions! This would also make the code quite unreadable as you could have a tiny function of 2 lines with 2 arrays that then needs an extra 2*10 lines to deallocate those arrays.

> we can always call the Fortran `dealloc` which will do nothing if the array doesn't belong to it

How would you do this. Presumably your "Fortran `dealloc`" is not the `deallocate` function?

I'm not sure if your method deallocates local arrays

**Reply by saidctb** [reply]
Yes, the performance issue is not that bad, because generally the accelerated function is not used in a loop otherwise the loop will need to be accelerated also. The issue is more with the memory.
Even though in my computer, creating an array and making a copy takes `0.15` seconds for a NumPy array of size `10**8`, but this can always be solved by creating outside and passing it as an argument

**Reply by saidctb** [reply]
> > we can always call the Fortran `dealloc` which will do nothing if the array doesn't belong to it, and then call the numpy `dealloc` afterwards
> 
> We can but this could be really inefficient. If we do this then we have to call the `dealloc` function for all functions in the scope (either in the module or imported) which return arrays. Depending on the size of the project that could be a huge number of functions! This would also make the code quite unreadable as you could have a tiny function of 2 lines with 2 arrays that then needs an extra 2*10 lines to deallocate those arrays.
> 
> > we can always call the Fortran `dealloc` which will do nothing if the array doesn't belong to it
> 
> How would you do this. Presumably your "Fortran `dealloc`" is not the `deallocate` function?
> 
> I'm not sure if your method deallocates local arrays

we will have one module variable and one `dealloc` function in the module that manages all the arrays created by all the functions in the module.

**Reply by EmilyBourne** [reply]
>we will have one module variable and one dealloc function in the module that manages all the arrays created by all the functions in the module.

So in the following:
```
def sum_to_n(n : int):
    a = np.arange(n)
    res = 0.0
    for ai in a:
        res += ai
    return res
```

You would make `a` a module variable?

This greatly increases the chances of name collisions

**Reply by EmilyBourne** [reply]
I suspect that this is sufficiently far from usual practice that this will end up confusing compilers and removing any chance of it making sensible optimisations.

Also making all arrays into globals is definitely an anti-pattern.

I don't think a memory problem is enough to warrant adding this amount of ugliness. If the user has a memory problem then they can also refactor their code to make the array into an inout argument instead of a result.

**Reply by EmilyBourne** [reply]
> Yes, the performance issue is not that bad, because generally the accelerated function is not used in a loop otherwise the loop will need to be accelerated also. The issue is more with the memory.

But the 100 deallocations also need to be called in the accelerated loop don't they?

**Reply by saidctb** [reply]
> > we will have one module variable and one dealloc function in the module that manages all the arrays created by all the functions in the module.
> 
> So in the following:
> 
> ```
> def sum_to_n(n : int):
>     a = np.arange(n)
>     res = 0.0
>     for ai in a:
>         res += ai
>     return res
> ```
> 
> You would make `a` a module variable?
> 
> This greatly increases the chances of name collisions
we will have one module variable that stores all the created arrays that are going to be returned by a function, so I don't see the issue of the name collision.
In your example, nothing will be stored because it's a variable that will not be returned

**Reply by EmilyBourne** [reply]
Ok. So then we do need to keep track of where it was allocated. As the deallocate statements for `a` and `b` in the following examples would be different:
```python
def sum_to_n(n : int):
    a = np.arange(n)
    b = my_range(n)
    res = 0.0
    for ai,bi in zip(a,b):
        res += ai*bi
    return res
```

**Reply by saidctb** [reply]
> But the 100 deallocations also need to be called in the accelerated loop don't they?
This is not a problem because it will not call the deallocate function that we defined in the `bind_c` files, the accelerated loop will be on the main Fortran Module, so the deallocation is handled with what we have now.

The proposed solution will be completely safe from name collisions or any possible complications, and will improve the performance for sure, the only thing that we need to add in each `bind_c` module is a `dealloc` function and one private `variable` module that keeps track of all the variables created in the `bind_c` files that are returned.

**Reply by saidctb** [reply]
> Ok. So then we do need to keep track of where it was allocated. As the deallocate statements for `a` and `b` in the following examples would be different:
> 
> ```python
> def sum_to_n(n : int):
>     a = np.arange(n)
>     b = my_range(n)
>     res = 0.0
>     for ai,bi in zip(a,b):
>         res += ai*bi
>     return res
> ```

is your function accelerated or not?
if it's accelerated then we do nothing, if it's not accelerated then yes  the `b` object will have a different `dealloc` then `a`

**Reply by saidctb** [reply]
maybe it's not clear yet, if you want we can have a quick meeting to talk about this

**Reply by EmilyBourne** [reply]
If you think it's useful then we can have a meeting. My function is accelerated.

Is your dictionary in the `bind_c` module? This would explain my confusion

**Reply by saidctb** [reply]
yes, it's only in the bind c module, so we will have no issue with your example, it will be treated as it is right now with the current Pyccel version

**Reply by saidctb** [reply]
For me, this solution is not a priority, I prefer to use the other one first, and then we can think about having this one if we want to have the same behavior as the other one

**Comment by EmilyBourne** [comment]
Given that we will only have this restriction in Fortran I don't think this restriction is a huge problem. f2py also did lots of copies internally.

It would be nice to be able to use the new standard though. Either we can wait until it is supported by PGI, or we could add a standard flag to pyccel. In this case we would also be able to use other new things from different standards. I don't know if there is anything useful that we don't use currently due to standard restrictions? But if we want to add C++ one day I imagine this flag could be useful

**Reply by saidctb** [reply]
We have this problem with OpenMP and OpenACC versions also, different compilers support different versions

**Comment by EmilyBourne** [comment]
I wonder if doing (Fortran 2003):
```fortran
  !........................................
  subroutine bind_c_f(n, bound_a, Dummy_0000, Dummy_0001) bind(c)

    implicit none

    integer(i64), value :: n
    type(c_ptr), intent(out) :: bound_a
    integer(i64), intent(out) :: Dummy_0000
    integer(i64), intent(out) :: Dummy_0001
    real(f64), pointer :: a_ptr(:,:)
    real(f64), allocatable :: a(:,:)

    call f(n, a = a)
    Dummy_0000 = size(a, 2_i64, i64)
    Dummy_0001 = size(a, 1_i64, i64)
    allocate(a_ptr(0:size(a, 1_i64, i64) - 1_i64, 0:size(a, 2_i64, i64) &
          - 1_i64), source_expr = a)
    bound_a = c_loc(a_ptr)

  end subroutine bind_c_f
```
Instead of:
```
  !........................................
  subroutine bind_c_f(n, bound_a, Dummy_0000, Dummy_0001) bind(c)

    implicit none

    integer(i64), value :: n
    type(c_ptr), intent(out) :: bound_a
    integer(i64), intent(out) :: Dummy_0000
    integer(i64), intent(out) :: Dummy_0001
    real(f64), pointer :: a_ptr(:,:)
    real(f64), allocatable :: a(:,:)

    call f(n, a = a)
    Dummy_0000 = size(a, 2_i64, i64)
    Dummy_0001 = size(a, 1_i64, i64)
    allocate(a_ptr(0:size(a, 1_i64, i64) - 1_i64, 0:size(a, 2_i64, i64) &
          - 1_i64))
    a_ptr = a
    bound_a = c_loc(a_ptr)

  end subroutine bind_c_f
```
might help the compiler realise that `a` is not used again and the alloc could be optimised to a move

**Reply by saidctb** [reply]
Using `source_expr` can be an improvement for sure

**Reply by EmilyBourne** [reply]
@saidctb would you be able to test this change manually with the large test case you are using to see if it changes anything?

**Reply by EmilyBourne** [reply]
@saidctb not sure if you saw my comment as we replied at the same time. Can you test if this lets the compiler optimise out the extra memory?

**Reply by saidctb** [reply]
ah yes, I can do this

**Reply by saidctb** [reply]
Fortran gives me this error
```fortran
       allocate( data_array(size) , source_expr=temp_array)
                                   1
Error: Allocate-object at (1) is neither a data pointer nor an allocatable variable
```

**Reply by EmilyBourne** [reply]
What about:
```
allocate( data_array , source_expr=temp_array)
```
I'm not sure if the size parameter is necessary/allowed, I just left it in to try and get the right lbound (but that's not necessary here)

**Reply by saidctb** [reply]
The same error

**Reply by saidctb** [reply]
```fortran
module dealloc_test

   use, intrinsic :: iso_c_binding, only : c_double, c_ptr, c_size_t, c_associated, c_loc

   implicit none

   real(c_double), pointer, protected, save :: data_array(:) ! Named, protected target data object

contains

   function alloc(length) bind(C, name='alloc') result(rv)

      ! Argument list
      integer(c_size_t), value, intent(in) :: length
      ! Function result
      type (c_ptr) :: rv
      real(c_double), allocatable :: temp_array(:)

      allocate( temp_array(length) )
      ! Handling elided for an existing allocation of data_array
      allocate( data_array(length) , source_expr=temp_array)

      rv = c_loc(data_array)

      return

   end function

   subroutine dealloc(ptr) bind(C, name='dealloc')

      ! Argument list
      type(c_ptr), value, intent(in) :: ptr

      ! Check if input pointer is associated with module variable
      if ( c_associated(ptr, c_loc(data_array)) ) then
         deallocate( data_array )
      end if

      return

   end subroutine

end module
```
this is the Fortran file you can try to compile it

**Reply by EmilyBourne** [reply]
IBM has misleading docs. The keyword is `source` not `source_expr`. I'm not sure why you need a module pointer here though? Are you trying to test this suggestion and your dictionary suggestion at the same time?

**Reply by saidctb** [reply]
It compiles using `source`

**Reply by saidctb** [reply]
so using `source` it still creates a copy

**Reply by EmilyBourne** [reply]
Yes. Which is what it's supposed to do. The only function that I've found that doesn't create a copy is `move_alloc`
but that requires 2 allocatables otherwise it raises:
```
   24 |     call move_alloc(Out_0001, Out_0001_ptr)
      |                              1
Error: ‘to’ argument of ‘move_alloc’ intrinsic at (1) must be ALLOCATABLE
```
I was just hoping that the compiler might optimise out the copy :(

**Reply by saidctb** [reply]
yes, sadly we can't do much with what we have

**Comment by EmilyBourne** [comment]
@saidctb what code were you using to test how big an array can be returned?

**Reply by saidctb** [reply]
I'm just testing it in C

**Reply by saidctb** [reply]
```C
#include <stdio.h>

/*gcc -c test_save.c */
/*gfortran test_save.o test_save.f90 -o exe*/

typedef void * FDATA; // An opaque pointer to some "data" in Fortran

FDATA alloc( size_t );
void dealloc( FDATA );

int main() {

   FDATA dat1,dat2;
   size_t n1 = 2000000000;
/*   size_t n2 = 11;*/

   dat1 = alloc(n1);
/*   dat2 = alloc(n2);*/
   // Employing "domain knowledge" data in Fortran points to a "double array"!
   double *array1 = (double *)dat1;
/*   double *array2 = (double *)dat2;*/

   printf("C main: array1 = \n");
/*   for (size_t i=0; i < n1; i++) {*/
/*      printf("%lf ", array1[i]);*/
/*   }*/
/*   printf("\n ");*/

/*   printf("C main: array2 = \n");*/
/*   for (size_t i=0; i < n2; i++) {*/
/*      printf("%lf ", array2[i]);*/
/*   }*/
/*   printf("\n ");*/
   dealloc(dat1);
/*   dealloc(dat2);*/

   return 0;
}
```
This is the C file
you can compile and run the files using
```
gcc -c test_save.c
gfortran test_save.o test_save.f90 -o exe
./exe
```

**Reply by EmilyBourne** [reply]
With the test cases:
```python
import sys
from numpy import ones

n = int(sys.argv[1])

a = ones(n)
print(a)
```
and
```python
import sys
from tmp import f

n = int(sys.argv[1])

a = f(n)
print(a)
```
with the accelerated file tmp.py:
```python
import numpy as np

def f(n :int):
    return np.ones(n)
```

I can create 2x bigger arrays with numpy. This doesn't change if I change the `allocate` expression

**Reply by saidctb** [reply]
our computers are different so the limit is going to be different, at a certain point, you can't go bigger in your computer with NumPy arrays, it should be half the size using `allocate`

**Reply by EmilyBourne** [reply]
That is what I am seeing n=500000000 works for numpy, but not for pyccel, n=250000000 works for both

**Reply by saidctb** [reply]
Yes, makes sense
